### PR TITLE
Fixed  discovery weight bug

### PR DESCRIPTION
**Pull request by frank0417**
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible?

### CODE REVIEW

**Comment by spacewander** [comment]
@frank0417 
Please explain why the 'fix' is needed.

**Comment by frank0417** [comment]
> @frank0417
> Please explain why the fix is needed, and add test case for it.

When Eureka is used as a service discovery, the service is configured with weights. For example, the weight of servicea1 is configured as metadata- map.weight:20 ，ServiceA2 The weight configuration of is: metadata- map.weight:80 At present, the weight value of apifix is cached according to the string when parsing, which results in the request can not be handled correctly when the load is loaded, and the load balancing can not be carried out according to the weight set by the service. The default weight is used for the load, for example, the weight is 100

**Comment by frank0417** [comment]
> @frank0417
> Please explain why the fix is needed, and add test case for it.
test case:
ServiceA1:
server:
  port: 9097

eureka:
  instance:
    leaseRenewalIntervalInSeconds: 1
    leaseExpirationDurationInSeconds: 2
    metadata-map:
      weight: 20
  client:
    serviceUrl:
      defaultZone: http://127.0.0.1:8761/eureka/
    healthcheck:
      enabled: true
    lease:
      duration: 5


serviceA2 :

  server:
  port: 9098

eureka:
  instance:
    leaseRenewalIntervalInSeconds: 1
    leaseExpirationDurationInSeconds: 2
    metadata-map:
      weight: 80
  client:
    serviceUrl:
      defaultZone: http://127.0.0.1:8761/eureka/
    healthcheck:
      enabled: true
    lease:
      duration: 5

spring:
  application:
    name: customer-service

**Comment by membphis** [comment]
@qiujiayu please take a look at this PR, need you review

File: apisix/discovery/eureka.lua [hunk 1]
```diff
@@ -197,22 +197,26 @@ local function fetch_full_registry(premature)
...
         for _, instance in ipairs(app.instance) do
             local ip, port, metadata = parse_instance(instance)
+        
             if ip and port then
                 local nodes = up_apps[app.name]
                 if not nodes then
                     nodes = core.table.new(#app.instance, 0)
                     up_apps[app.name] = nodes
                 end
+                local tmp_weight = default_weight
```
Comment (qiujiayu) [inline apisix/discovery/eureka.lua:207]: It would be better to use 'weight'

**Comment by membphis** [comment]
@frank0417 please take a look at the output of CI: https://github.com/apache/apisix/pull/1989/checks?check_run_id=952820522#step:6:125

you need to fix them first

File: apisix/discovery/eureka.lua [hunk 2]
```diff
@@ -34,6 +35,8 @@ local log                = core.log
 local default_weight
 local applications
 
+
+
```
Comment (membphis) [inline apisix/discovery/eureka.lua:39]: two blank lines are enough but got `3`

File: apisix/discovery/eureka.lua [hunk 3]
```diff
@@ -203,16 +205,18 @@ local function fetch_full_registry(premature)
                     nodes = core.table.new(#app.instance, 0)
                     up_apps[app.name] = nodes
                 end
+                local weight = default_weight
+                if metadata and metadata.weight then
+                    weight = tonumber(metadata.weight)
+                else
+                    metadata.weight = nil
+                end
```
[the remark this answers is not in the document]
Comment (frank0417) [reply apisix/discovery/eureka.lua:213]: ok，the result of tonumber may be nil

File: apisix/discovery/eureka.lua [hunk 4]
```diff
@@ -203,16 +205,18 @@ local function fetch_full_registry(premature)
                     nodes = core.table.new(#app.instance, 0)
                     up_apps[app.name] = nodes
                 end
+                local weight = default_weight
+                if metadata and metadata.weight then
+                    weight = tonumber(metadata.weight) or default_weight
+                else
+                    metadata.weight = nil
```
Comment (membphis) [inline apisix/discovery/eureka.lua:212]: why we need this line? Can we remove this line?

Comment (membphis) [reply apisix/discovery/eureka.lua:212]: and we need test case to confirm this change is correct.

Comment (frank0417) [reply apisix/discovery/eureka.lua:212]: > why we need this line? Can we remove this line?

Which line ？

Comment (membphis) [reply apisix/discovery/eureka.lua:212]: `metadata.weight = nil`

Comment (frank0417) [reply apisix/discovery/eureka.lua:212]: > and we need test case to confirm this change is correct.

We need to learn about the test framework. We haven't touched it before. We will submit the test case later.
haha

**Comment by frank0417** [comment]
> missing test case

test cases have been added

File: apisix/discovery/eureka.lua (hunk 4 above)
Comment (frank0417) [reply apisix/discovery/eureka.lua:212]: > `metadata.weight = nil`
   remove useless data

File: t/discovery/eureka-weight.t [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+--- request
+GET /t
+--- yaml_config eval: $::yaml_config
+--- response_body
+[{"count":8,"port":"1981"},{"count":2,"port":"1980"}]
+--- no_error_log
+[error]
+
+
+
```
Comment (membphis) [inline t/discovery/eureka-weight.t:158]: only one blank line is need.

Comment (frank0417) [reply t/discovery/eureka-weight.t:158]: Test cases can run through the local test case, and fail to test on CI. Can you give me some advice

Comment (membphis) [reply t/discovery/eureka-weight.t:158]: > Test cases can run through the local test case, and fail to test on CI. Can you give me some advice

@qiujiayu need your help ^_^

**Comment by moonming** [comment]
@frank0417 @membphis ping

**Pull request closed without merging**